### ISSUE TITLE
Sending Service downtimes locks up database until 400K queries are pending 

### ISSUE THREAD

**Issue by towolf** [Icinga/icinga2#5080]
Continued from https://github.com/Icinga/icingaweb2/issues/2761

*relevant screenshot after downtiming 1000 services ...*
![image](https://cloud.githubusercontent.com/assets/345210/24043088/1fe99c56-0b15-11e7-8286-54a0b5ff3d8d.png)

@Thomas-Gelf here's the EXPLAIN for the downtimes

```
EXPLAIN SELECT downtimehistory_id   FROM icinga_downtimehistory   WHERE entry_time = FROM_UNIXTIME(1489674017)   AND instance_id = 1   AND internal_downtime_id = '792'   AND scheduled_end_time = FROM_UNIXTIME(1489   AND scheduled_start_time = FROM_UNIXTIME(1489673972)\G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: icinga_downtimehistory
   partitions: NULL
         type: ref
possible_keys: instance_id,instance_id_debug
          key: instance_id_debug
      key_len: 23
          ref: const,const,const
         rows: 1
     filtered: 5.00
        Extra: Using index condition; Using where
```

And 

```
SELECT COUNT(*) FROM icinga_downtimehistory;
+----------+
| COUNT(*) |
+----------+
|   399983 |
+----------+
```

**Comment by hrld** [comment]
c&p from the other issue:

There was no useful index. Should there be more than these?

PRIMARY KEY (downtimehistory_id),
UNIQUE KEY instance_id (instance_id,object_id,entry_time,internal_downtime_id),
KEY sla_idx_dohist (object_id,actual_start_time,actual_end_time)

instance_id always 1, object_id not in query. Index not helpful.
With an index on icinga_downtimehistory(instance_id,internal_downtime_id,entry_time) the performance problem is solved.

**Comment by towolf** [comment]
So, did you "fix it" locally @hrld? Seems like I went home too early ...

**Comment by dnsmichi** [comment]
Seems TriggerDowntime() uses a where clause which does not match the index.

https://github.com/Icinga/icinga2/blob/master/lib/db_ido/dbevents.cpp#L653

**Comment by Thomas-Gelf** [comment]
@towolf: The EXPLAIN doesn't match the shown keys, instance_id_debug is new to me and not shown in your schema snippet. However, doesn't look that bad after all. According to your EXPLAIN, the query should be fast. A DB server shouldn't get nervous with 400k rows using a 23 byte compound key. Did you tune your database or is this a default configuration? What kind of DB are you using, how does your setup look like? Any kind of replication? What kind of hardware/storage are you using?

You might do some research whether this is really useful data you want to preserve, cleaning up that table might help. Check whether there are some anomalies or objects going wild by running queries like this one:

```sql
SELECT object_id, COUNT(*) AS cnt, DATE_FORMAT(entry_time, '%Y-%m %U') AS week
  FROM icinga_downtimehistory GROUP BY object_id, week ORDER BY week, cnt;
```

I would cleanup that table in case it had millions of rows, but (I know I repeat myself) 400k shouldn't hurt. One second per query when running 1000 queries is way too much, 1000 per second would make more sense to me.

@dnsmichi: Not sure how those queries are executed, there is one possible issue I could imagine: @towolf mentioned that he injects 1000 downtimes with a single API request and a single filter. Now in case our code does something like "START TRANSACTION; [while downtime: [check if matching downtime exists]; [insert downtime]; [end]; COMMIT;" - and you combine this with slow round-trips or some delay by async operation, a similar effect might occur. But that's just a completely blind guess.

**Comment by Thomas-Gelf** [comment]
@hrld: sorry, I misread your post:

> With an index on icinga_downtimehistory(instance_id,internal_downtime_id,entry_time) the performance problem is solved.

Cool, great to hear that!

**Comment by towolf** [comment]
@Thomas-Gelf 

It's not a super beefy server, but from all that we could see the queries were either repetitive and complex and did not benefit from indices (the icingaweb auto-refresh) or they were downright DOS-like and harmful and had weird unhelpful indices (like the downtime services thing).

So my approach was, there is too much work, let's reduce that work. And hence my 2 PRs.

But apparently my colleagues found the missing index after I left work ... so all is good. 

The PRs still deserve some thought I think.

**Comment by dnsmichi** [comment]
@Thomas-Gelf I haven't seen the exact API call to create those downtimes, but I would believe that either 1) it is a fixed downtime started immediately or 2) a flexible downtime triggered because objects are in a NOT-OK state.

It shouldn't matter if you schedule 1000 service or host downtimes, the database event is the same. If you've got the downtimehistory table with many entries and no cleanup, a broken index, that should be fairly easy to reproduce. Thing is, my Icinga 2 on my macbook doesn't have such a long history to run into the issue while testing other things.

There was a fix a while ago to add the object_id for downtime removal lockup. https://github.com/Icinga/icinga2/blame/master/lib/db_ido/dbevents.cpp#L613 - #3383

I believe that's missing for triggering the downtimes too, so the where condition matches a partial index.

**Comment by Thomas-Gelf** [comment]
> It's not a super beefy server [..]

Well, in case the index doesn't fit you're lost anyways - so it's good that we found this.

> The PRs still deserve some thought I think.

Both of them try to tackle a real world problem, they are absolutely worth to be dedicated some more time to refine them.

Thanks a lot for your time and your valuable input!

Cheers,
Thomas

**Comment by towolf** [comment]
@dnsmichi, it's `1) it is a fixed downtime started immediately`

This is a python script we wrote for various tasks, starting maintenances for a whole project, quickly rebooting a server, etc.

The API calls looks like this:

```
{
  "comment": "icingatest",
  "filter": "\"my_product\" in host.vars.product",
  "end_time": 1489789196.186792,
  "author": "tobias",
  "duration": 200,
  "start_time": 1489788988.186792,
  "fixed": true,
  "type": "Host"
}
Request URL: https://our-icinga-master:5665/v1/actions/schedule-downtime
```
and the same for `"type": "Service"`

The `"type": "Host"` takes a few seconds, the `"type": "Service"` call 3 minutes.

Only, before adding the proper index, the db went into lockup for 15 minutes and now it does not even register as out-of-control "Last Status Update":

![image](https://cloud.githubusercontent.com/assets/345210/24065040/7c022fb4-0b68-11e7-8a7a-089a7d231c7a.png)

**Comment by towolf** [comment]
And I don't think anybody has ever looked at the downtime history, at least not for anything longer than the current day.

**Comment by dnsmichi** [comment]
Since this affects multiple (customer) setups we decided to backport the fix for 2.6.3.
